File: uva/100-the-3n-plus-1-problem/c/main.c
#include <stdio.h>

#define MAX_INPUT 1000000
#define MAX(x,y) x > y ? x : y
#define MIN(x,y) x < y ? x : y
#define Si(n) scanf("%d", &n)

int lengths[MAX_INPUT];

int collatz_len(const int n) {
  int len;
  if(n < MAX_INPUT && lengths[n] > 0)
    return lengths[n];
  if(n & 1)
    len = collatz_len(3*n + 1) + 1;
  else
    len = collatz_len(n >> 1) + 1;
  if(n < MAX_INPUT)
    lengths[n] = len;
  return len;
}

int main() {
  int i, j, ii, jj, max;
  lengths[1] = 1;
  for(i = 2; i<MAX_INPUT; i++) lengths[i] = 0;
  while(Si(i) != EOF) {
    Si(j);
    ii = MIN(i, j);
    jj = MAX(i, j);
    for(max = 0; ii<=jj; ii++)
      max = MAX(max, collatz_len(ii));
    printf("%d %d %d\n", i, j, max);
  }
  return 0;
}

File: uva/100-the-3n-plus-1-problem/python/main.py
#!/usr/bin/env python

# For the set of natural numbers between X and Y inclusive, report the
# longest collatz sequence.

import re

def min(x,y):
    if x < y:
        return x
    else:
        return y

def max(x,y):
    if x < y:
        return y
    else:
        return x

def collatz_len(n):
    if n == 1:
        return 1
    elif n & 1 == 1:
        return collatz_len(3*n+1) + 1
    else:
        return collatz_len(n/2) + 1

with open('/dev/stdin', 'rt') as f:
    line = f.readline()
    while line:
        i, j = map(int, re.split(r' +', line))
        ii = min(i,j)
        jj = max(i,j)
        m = 0
        while ii <= jj:
            m = max(m, collatz_len(ii))
            ii += 1
        print i, j, m
        line = f.readline()
